* Panmethyl

Panmethyl is a nextflow pipeline maps methylation data from long-reads to pangenomes.

** Apptainer image
#+begin_src shell
apptainer pull --arch amd64 panmethyl.sif library://cgroza/collection/panmethyl:latest
#+end_src

** Inputs
It takes the following inputs:

+ ~--out~ - directory in which panmethyl will write the output files.
+ ~--bams~ - CSV file listing the BAM files to be mapped to the pangenome.
The format of this CSV file is:

#+begin_src shell
  sample,path
  name1,path/to/bam1
  name2,path/to/bam2
#+end_src

 These BAM files must be annotated with the appropriate methylation information.
 For example, the location of modified bases must be encoded in the ~MM~ tag
 and the likelihoods of methylation must be encoded in the ~ML~ tag.

+ ~--graph~ - Pangenome in GFA format.

** Outputs

Panmethyl outputs a CSV plain text file for each entry in
~--bams~. This file is a CSV file listing the graph node, the position of the
modified base, its strand, the coverage on the modified base, and the average
methylation level, encoded on a scale from 0 to 255 (as in the ML tag).

** Citation
If using panmethyl, please cite:
- Groza, C., Ge, B., Cheung, W. A., Pastinen, T., & Bourque, G. (2025). Expanded methylome and quantitative trait loci detection by long-read profiling of personal DNA. Genome research, 35(4), 644â€“652. https://doi.org/10.1101/gr.279240.124
